pkg/operator/target_status.go

```diff
@@ -0,0 +1,936 @@
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/api/option"
+	apihttp "google.golang.org/api/transport/http"
+	"google.golang.org/grpc"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// How often to do a full refresh of cached Kubernetes data.
+	informerResyncInterval = time.Minute * 15
+	// How many targets to keep in each group.
+	maxSampleTargetSize = 5
+)
+
+var (
+	targetStatusDuration = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "prometheus_engine_target_status_duration",
+		Help: "A metric indicating how long it took to fetch the complete target status.",
+	}, []string{})
+)
+
+// Wraps a time.Ticker so we can replace for unit testing.
+type Ticker interface {
+	// Stops the ticker from ticking.
+	Stop()
+	// Returns the channel receiving ticks.
+	Channel() <-chan time.Time
+}
+
+type timeTicker time.Timer
+
+func NewTicker(d time.Duration) Ticker {
+	data := timeTicker(*time.NewTimer(d))
+	return &data
+}
+
+func (t *timeTicker) Stop() {
+	ticker := time.Timer(*t)
+	ticker.Stop()
+}
+
+func (t *timeTicker) Channel() <-chan time.Time {
+	ticker := time.Timer(*t)
+	return ticker.C
+}
+
+type TargetStatusProcessor interface {
+	Start(ctx context.Context, tickerBuilder func() Ticker) error
+}
+
+type dataManagerBuilder interface {
+	build() (dataManager, error)
+}
+
+// Manages fetching and patching Kubernetes data.
+type dataManager interface {
+	// Returns all pods running Prometheus, mapped with port as the key.
+	getPrometheusPods(ctx context.Context) (map[int32][]*corev1.Pod, error)
+
+	getPodMonitorings(ctx context.Context) ([]monitoringv1.PodMonitoring, error)
+	getClusterPodMonitorings(ctx context.Context) ([]monitoringv1.ClusterPodMonitoring, error)
+
+	patchPodMonitoring(ctx context.Context, podMonitoring *monitoringv1.PodMonitoring) error
+	patchClusterPodMonitoring(ctx context.Context, podMonitoring *monitoringv1.ClusterPodMonitoring) error
+
+	// Returns the target fetcher, with an zero-index number `thread` that signifies which goroutine this is for.
+	getTargetFetcher(ctx context.Context, thread uint16) (targetFetcher, error)
```

Comment:
I made it a `targetFetcherBuilder` as opposed to a pure function because there are some lookups that only need to be done once as opposed to per-thread.

Reply:
It seems like credentials are not necessary but we still save and re-use the client for all threads.